src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+            VirtualPosListCmpContext{reference_matrix_right});
+
+  /**
+   * Build result table
+   */
+  auto left_idx = size_t{0};
+  auto right_idx = size_t{0};
+  const auto num_left = virtual_pos_list_left.size();
+  const auto num_right = virtual_pos_list_right.size();
+
+  // Somewhat random way to decide on a chunk size.
+  const auto out_chunk_size = std::max(_input_table_left()->chunk_size(), _input_table_right()->chunk_size());
+
+  auto out_table = Table::create_with_layout_from(_input_table_left(), out_chunk_size);
+
+  std::vector<std::shared_ptr<PosList>> pos_lists(reference_matrix_left.size());
+  std::generate(pos_lists.begin(), pos_lists.end(), [&] { return std::make_shared<PosList>(); });
+
+  // Adds one row to the pos_lists currently being filled
+  const auto emit_row = [&](const ReferenceMatrix& reference_matrix, size_t row_idx) {
+    for (size_t pos_list_idx = 0; pos_list_idx < pos_lists.size(); ++pos_list_idx) {
+      pos_lists[pos_list_idx]->emplace_back(reference_matrix[pos_list_idx][row_idx]);
+    }
+  };
+
+  // Turn 'pos_lists' into a new chunk and append it to the table
+  const auto emit_chunk = [&]() {
+    Chunk chunk;
+
+    for (size_t pos_lists_idx = 0; pos_lists_idx < pos_lists.size(); ++pos_lists_idx) {
+      const auto segment_column_id_begin = _column_segment_begins[pos_lists_idx];
+      const auto segment_column_id_end = pos_lists_idx >= _column_segment_begins.size() - 1
+                                             ? _input_table_left()->column_count()
+                                             : _column_segment_begins[pos_lists_idx + 1];
+      for (auto column_id = segment_column_id_begin; column_id < segment_column_id_end; ++column_id) {
+        auto ref_column = std::make_shared<ReferenceColumn>(
+            _referenced_tables[pos_lists_idx], _referenced_column_ids[column_id], pos_lists[pos_lists_idx]);
+        chunk.add_column(ref_column);
+      }
+    }
+
+    out_table->emplace_chunk(std::move(chunk));
+  };
+
+  // Comparator of two ReferenceMatrix rows
+  const auto cmp = [](const auto& matrix_a, const auto idx_a, const auto& matrix_b, const auto idx_b) {
+    for (size_t column_idx = 0; column_idx < matrix_a.size(); ++column_idx) {
+      if (matrix_a[column_idx][idx_a] < matrix_b[column_idx][idx_b]) return true;
+      if (matrix_b[column_idx][idx_b] < matrix_a[column_idx][idx_a]) return false;
+    }
+    return false;
+  };
+
+  size_t chunk_row_idx = 0;
+  for (; left_idx < num_left || right_idx < num_right;) {
+    /**
+     * Begin derived from std::union()
+     */
+    if (left_idx == num_left) {
+      emit_row(reference_matrix_right, virtual_pos_list_right[right_idx]);
+      ++right_idx;
+    } else if (right_idx == num_right) {
+      emit_row(reference_matrix_left, virtual_pos_list_left[left_idx]);
+      ++left_idx;
+    } else if (cmp(reference_matrix_right, virtual_pos_list_right[right_idx], reference_matrix_left,
```

Comment:
I find it hard to follow when left and right switch. Why does right come first here?

Reply:
Because I don't want two implement a second comparator (i.e. operator>())

Reply:
valid